### PR TITLE
dmctl: show help msg before validate cfg #1564

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve? 

#1564 

### What is changed and how it works?

dmctl：print help msg when user input `--help` before validate the dmctl config

### Check List 

Tests 
 - Unit test

Code changes
- `cmd/dm-ctl/main.go`
- `dm/ctl/common/config.go `

Side effects
- None

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### CODE REVIEW

File: cmd/dm-ctl/main.go [hunk 1]
```diff
@@ -119,7 +119,18 @@ func main() {
...
-		common.PrintLinesf("parse cmd flags err: %s", terror.Message(err))
+		// NOTE:  when `--help` in cmdArgs we need to print out the help msg.)
+		var needPrintHelp bool
+		for _, cmd := range cmdArgs {
+			if cmd == "--help" {
+				needPrintHelp = true
+			}
+		}
+		if needPrintHelp {
+			ctl.PrintHelp(cmdArgs)
```
Comment (lichunzhu) [inline cmd/dm-ctl/main.go:130]:
```suggestion
			ctl.PrintHelp(cmdArgs)
			os.Exit(0)
```

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

**Review by lichunzhu — COMMENTED** [review]
Could you please add a integration test in `tests/dmctl_basic`? Check the return code there.

**Review by lichunzhu — COMMENTED** [review]
/lgtm

File: cmd/dm-ctl/main.go [hunk 2]
```diff
@@ -119,8 +119,22 @@ func main() {
 		}
 		os.Exit(0)
 	default:
-		common.PrintLinesf("parse cmd flags err: %s", terror.Message(err))
-		os.Exit(2)
+		// NOTE:  when `--help` in cmdArgs we need to print out the help msg.)
```
Comment (lance6716) [inline cmd/dm-ctl/main.go:122]:
```suggestion
		// NOTE: when `--help` in cmdArgs we need to print out the help msg :)
```

an emoji?

File: cmd/dm-ctl/main.go [hunk 3]
```diff
@@ -119,8 +119,22 @@ func main() {
...
+		var needPrintHelp bool
+		for _, cmd := range cmdArgs {
+			if cmd == "--help" {
+				needPrintHelp = true
```
Comment (lance6716) [inline cmd/dm-ctl/main.go:126]: seems `ctl.PrintHelp(cmdArgs)` and `os.Exit(0)` could be moved here. Also we don't need `needPrintHelp`

File: tests/dmctl_command/run.sh [hunk 4]
```diff
@@ -47,6 +47,22 @@ function run() {
         exit 1
     fi
 
+
+    # check dmctl command start-task --help
+    # it should print the usage for start-task
+    $PWD/bin/dmctl.test DEVEL start-task --help > $WORK_DIR/help.log
+    help_msg=$(cat $WORK_DIR/help.log)
+    help_msg_cnt=$(echo "${help_msg}" | grep  "Usage:" | wc -l |xargs)
```
Comment (lance6716) [inline tests/dmctl_command/run.sh:55]: this `xargs` is not used?

File: tests/dmctl_command/run.sh [hunk 5]
```diff
@@ -47,6 +47,22 @@ function run() {
...
+    if [ "$help_msg_cnt" != 1 ]; then
+        echo "dmctl case 4 help failed: $help_msg"
+        exit 1
+    fi
+    echo $help_msg | grep -q "Usage:"
```
Comment (lance6716) [inline tests/dmctl_command/run.sh:60]: I think this is duplicated with above check

**Comment by Ehco1996** [comment]
@lance6716 all comments have be resoved. PTAL :)

File: cmd/dm-ctl/main.go [hunk 6]
```diff
@@ -119,22 +119,15 @@ func main() {
 		}
 		os.Exit(0)
 	default:
-		// NOTE:  when `--help` in cmdArgs we need to print out the help msg.)
-		var needPrintHelp bool
+		// NOTE: when `--help` in cmdArgs we need to print out the help msg.
```
Comment (lance6716) [inline cmd/dm-ctl/main.go:122]: currently we will print help message for subcommand, so this checking is too loose

```shell
$ bin/dmctl.test DEVEL --master-addr=127.0.0.1:8261 show-ddl-locks --help

Shows un-resolved DDL locks.

Usage:
  dmctl show-ddl-locks [-s source ...] [task-name | task-file] [flags]

Flags:
  -h, --help   help for show-ddl-locks

Global Flags:
  -s, --source strings   MySQL Source ID.
```

**Review by lance6716 — APPROVED** [review]

**Pull request merged**